### PR TITLE
WIP for working with the Arista eAPI system

### PR DESCRIPTION
**Pull request by i3149**
Ads in metrics around:

`show mlag detail` and `show ip bgp summary vrf all` router commands. 

These get mapped to `kentik.eapi.mlag.` and `kentik.eapi.bgp.` metrics, for example `kentik.eapi.mlag.PortsConfigured`

Enable by setting up a eapi section in the target device config:

```
    ext:
      eapi_config:
        username: $USERNAME
        password: $PASSWORD
        transport: https
        port: 443
```

For example. Then these metrics come in along with regular SNMP metrics.

### CODE REVIEW

File: pkg/inputs/snmp/x/arista/eapi.go [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+			dst.CustomMetrics["MsgReceived"] = kt.MetricInfo{Oid: "computed", Mib: "computed", Profile: "eapi.bgp", Type: "eapi.bgp"}
+
+			dst.CustomBigInt["PrefixAccepted"] = int64(state.PrefixAccepted)
+			dst.CustomMetrics["PrefixAccepted"] = kt.MetricInfo{Oid: "computed", Mib: "computed", Profile: "eapi.bgp", Type: "eapi.bgp"}
+
+			dst.CustomBigInt["PeerState"] = stateMap[state.PeerState]
+			dst.CustomMetrics["PeerState"] = kt.MetricInfo{Oid: "computed", Mib: "computed", Profile: "eapi.bgp", Type: "eapi.bgp"}
+
+			dst.CustomBigInt["OutMsgQueue"] = int64(state.OutMsgQueue)
+			dst.CustomMetrics["OutMsgQueue"] = kt.MetricInfo{Oid: "computed", Mib: "computed", Profile: "eapi.bgp", Type: "eapi.bgp"}
```
Comment (thezackm) [inline pkg/inputs/snmp/x/arista/eapi.go:172]: thoughts on changing this to `Oid: "eapi"` ?

File: pkg/inputs/snmp/x/arista/eapi.go [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+	dst.CustomMetrics["PortsActivePartial"] = kt.MetricInfo{Oid: "computed", Mib: "computed", Profile: "eapi.mlag", Type: "eapi.mlag"}
+
+	dst.CustomBigInt["PortsInactive"] = sv.MlagPorts.Inactive
+	dst.CustomMetrics["PortsInactive"] = kt.MetricInfo{Oid: "computed", Mib: "computed", Profile: "eapi.mlag", Type: "eapi.mlag"}
+
+	dst.CustomBigInt["PortsConfigured"] = sv.MlagPorts.Configured
+	dst.CustomMetrics["PortsConfigured"] = kt.MetricInfo{Oid: "computed", Mib: "computed", Profile: "eapi.mlag", Type: "eapi.mlag"}
+
+	dst.CustomBigInt["PortsActiveFull"] = sv.MlagPorts.ActiveFull
+	dst.CustomMetrics["PortsActiveFull"] = kt.MetricInfo{Oid: "computed", Mib: "computed", Profile: "eapi.mlag", Type: "eapi.mlag"}
```
Comment (thezackm) [inline pkg/inputs/snmp/x/arista/eapi.go:301]: same as above; should we change the `kt.MetricInfo` to have `Oid: "eapi"`?

File: pkg/inputs/snmp/x/arista/eapi.go [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+			dst.CustomBigInt = map[string]int64{}
+			dst.EventType = kt.KENTIK_EVENT_SNMP_DEV_METRIC
+			dst.Provider = c.conf.Provider
+			dst.DeviceName = c.conf.DeviceName
+			dst.SrcAddr = c.conf.DeviceIP
+			dst.Timestamp = time.Now().Unix()
+			dst.CustomMetrics = map[string]kt.MetricInfo{}
+
+			dst.CustomBigInt["MLAGShutdown"] = shutdownMap[config.Shutdown()]
+			dst.CustomMetrics["MLAGShutdown"] = kt.MetricInfo{Oid: "computed", Mib: "computed", Profile: "eapi", Type: "eapi"}
```
Comment (thezackm) [inline pkg/inputs/snmp/x/arista/eapi.go:331]: one more time 😄

File: pkg/kt/snmp.go [hunk 4]
```diff
@@ -133,6 +133,19 @@ type V3SNMPConfig struct {
 	origStr                  string
 }
 
+type EAPIConfig struct {
+	Host      string `yaml:"host"`
+	Username  string `yaml:"username"`
+	Password  string `yaml:"password"`
+	Transport string `yaml:"transport"`
+	Port      int    `yaml:"port"`
+}
```
Comment (thezackm) [inline pkg/kt/snmp.go:142]: will this section accept env vars and aws secrets?

File: pkg/inputs/snmp/x/arista/eapi.go (hunk 2 above)
Comment (i3149) [reply pkg/inputs/snmp/x/arista/eapi.go:301]: Good call.

File: pkg/kt/snmp.go (hunk 4 above)
Comment (i3149) [reply pkg/kt/snmp.go:142]: will add.

**Comment by i3149** [comment]
I think this is ready for consideration for merging?

**Comment by Mesverrum** [comment]
At this point the Arista data collection is pretty much hard coded, have we done any work to model out a "profile" for api's yet?

**Comment by i3149** [comment]
> At this point the Arista data collection is pretty much hard coded, have we done any work to model out a "profile" for api's yet?

Good q. I could imagine a profile having a list of sections the user could opt into using? Kinda like how mibs are set up now. For the Arista, sections would be BGP and MLAG right now, but could be expanded to others. For what metrics come from each section though, the statically typed nature of go makes it hard to pick what is used. Say what you will about snmp but at least here the results are all in a standard format.

File: pkg/inputs/snmp/x/arista/eapi.go [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+
+	mlag_strings = map[string]int64{
+		"consistent": 1,
+		"up":         1,
+		"connected":  1,
+		"active":     1,
+		"primary":    1,
+		"secondary":  2,
+		"disabled":   0,
+		"unknown":    3,
```
Comment (thezackm) [inline pkg/inputs/snmp/x/arista/eapi.go:99]: should these map values be unique for each entry?

Comment (i3149) [reply pkg/inputs/snmp/x/arista/eapi.go:99]: yeah, probably. they are used in different metrics but I think its just weird to have this collision here.

**Review by thezackm — DISMISSED** [review]
🚀

**Review by thezackm — APPROVED** [review]

**Pull request merged**